### PR TITLE
Document Protocol Buffers used in Anki

### PR DESCRIPTION
**Pull request by RumovZ**
I think the different Protobuf implementations and how they are utilised in Anki are one of the trickier parts of the codebase, so I tried to write down the gotchas I encountered over the time.
@dae, @hgiesel, please correct and add if you notice something. Especially with the TS side, I don't have that much experience.

### CODE REVIEW

File: docs/protobuf.md [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+and each has its own pecularities. This document highlights some aspects relevant
+to Anki and hopefully helps to avoid some common pitfalls.
+
+For information about Protobuf's types and syntax, please see the official [language guide](https://developers.google.com/protocol-buffers/docs/proto3).
+
+## General Notes
+
+### Names
+
+Generated code follows the naming conventions of the targetd language. So to access
```
Comment (dae) [inline docs/protobuf.md:13]:
```suggestion
Generated code follows the naming conventions of the targeted language. So to access
```

File: docs/protobuf.md [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+safe to make as part of a schema upgrade, because schema 11 (the targeted schema
+when choosing _Downgrade_), does not make use of Protobuf messages.
+
+### Field Numbers
+
+Field numbers larger than 15 need an additional byte to encode, so `repeated` fields
+should preferrably be assigned a number between 1 and 15. If a message contains
+`reserved` fields, this is usually to accomodate potential future `repeated` fields.
+
+## Implementation-specific Notes
```
Comment (dae) [inline docs/protobuf.md:89]:
```suggestion
## Implementation-Specific Notes
```

File: docs/protobuf.md [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+Anki uses [protobuf.js](https://protobufjs.github.io/protobuf.js/), which offers
+some documentation.
+
+-   If using a message `Foo` as a type, make sure not to use the generated interface
+    `IFoo` instead. Their definitions are very similar, but the interface requires
+    null checks for every field.
+
+### Rust
+
+Anki uses the [protobuf crate](https://docs.rs/protobuf/latest/protobuf/index.html).
```
Comment (dae) [inline docs/protobuf.md:112]: This is wrong, we use https://github.com/tokio-rs/prost

File: docs/protobuf.md [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+It's documentation is not particularly useful, but it doesn't have to: From within
```
Comment (dae) [inline docs/protobuf.md:113]:
```suggestion
Its documentation is not particularly useful, but it doesn't have to: From within
```

Does that apply to Prost as well?

File: docs/protobuf.md [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+`anki/rslib` run `cargo doc --open --document-private-items`. Inside the `pb` module
+you will find all generated Rust types and their implementations.
+
+-   Enum variants arrive as `i32`s and must be mapped to the generated variants
```
Comment (dae) [inline docs/protobuf.md:117]: For simple enums (not oneofs), you can call message.enum_field() instead of Enum::from_i32(message.enum_field)

**Comment by dae** [comment]
Hah, I just posted #2043 before I saw you posted this.

**Comment by RumovZ** [comment]
Cool! That seems like a considerable improvement over protobuf.js.

Given that I had already forgotten that we use prost and `message.enum_field()` I'm glad I wrote this document. 😅 

> Does that apply to Prost as well?

I'd say yes, if it comes to working with the generated code. Compare that to the Python implementation thoroughly explaining how the Python code will look like. But as I wrote, Rust has something better. 🙂

**Comment by dae** [comment]
Thanks Rumo, looks good. Henrik, please feel free to submit any follow-up changes if you think something needs changing.

**Pull request merged**